src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -2207,11 +2207,37 @@ void EmitAtomic(RegexNode node, RegexNode? subsequent)
                 Debug.Assert(node.Kind is RegexNodeKind.Atomic, $"Unexpected type: {node.Kind}");
                 Debug.Assert(node.ChildCount() == 1, $"Expected 1 child, found {node.ChildCount()}");
 
-                // Atomic simply outputs the code for the child, but it ensures that any done label left
-                // set by the child is reset to what it was prior to the node's processing.  That way,
-                // anything later that tries to jump back won't see labels set inside the atomic.
+                RegexNode child = node.Child(0);
+
+                if (!analysis.MayBacktrack(child))
```

Comment:
> instead of doing the check here should we do it in EmitNode similar to how you are doing it for the source generator?

The source generator has a structure that's like:
- Emit nodes that don't need a header or scope
- Emit header and scope
- Emit nodes that need a header and scope

The compiler is just:
- Emit nodes (since header/scope is visual/language construct)

To do the same thing in the compiler, we'd be splitting out a separate switch for no particular reason.  I contemplated still doing the check in EmitNode and just only calling EmitAtomic when needed, but that then breeds inconsistency from the handling of every other node type.

>  in the source generator should we always call EmitAtomic and have the check in there that redirects to EmitNode when the child may not backtrack?

Then we'd either end up adding a header comment / scope even when it's superfluous, or we'd need to change the signature / implementation to support passing in whether the method should output that info and duplicate it.

Reply:
I see, that's fine then. I was only nitpicking mainly for consistency sake between both code generators where we can, but I totally agree that this should only be done where it makes sense.